migrations/versions/0320_optimise_notifications.py

```diff
@@ -0,0 +1,79 @@
+        op.execute('DROP INDEX IF EXISTS ix_notifications_service_id')
+
+        # Create new composite indexes
+        # PLEASE NOTE: that if you create index on production you need to add concurrently to the create statement,
+        # however we are unable to do that inside a transaction like this upgrade method
+        ix_notifications_service_id_composite = """
+        CREATE INDEX ix_notifications_service_id_composite 
+            on notifications (service_id, notification_type, notification_status, created_at)
+        """
+        op.execute(ix_notifications_service_id_composite)
+
+        ix_notifications_notification_type_composite = """
+            CREATE INDEX ix_notifications_notification_type_composite 
+            on notifications (notification_type, notification_status, created_at)
+        """
+        op.execute(ix_notifications_notification_type_composite)
+        # DROP and CREATE all other indexes
+        op.execute('DROP INDEX IF EXISTS ix_notifications_service_id_created_at')
+        op.execute(
+            'CREATE INDEX ix_notifications_service_id_created_at ON NOTIFICATIONS (service_id, created_at)'
+        )
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_client_reference')
+        op.execute('CREATE INDEX ix_notifications_client_reference on notifications (client_reference)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_client_reference')
+        op.execute('CREATE INDEX ix_notifications_client_reference ON notifications (client_reference)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_created_at')
+        op.execute('CREATE INDEX ix_notifications_created_at ON notifications (created_at)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_job_id')
+        op.execute('CREATE INDEX ix_notifications_job_id ON notifications (job_id)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_reference')
+        op.execute('CREATE INDEX ix_notifications_reference ON notifications (reference)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_service_created_at')
+        op.execute(
+            'CREATE INDEX ix_notifications_service_created_at ON notifications (service_id, created_at)')
+
+        op.execute('DROP INDEX IF EXISTS ix_notifications_template_id')
+        op.execute('CREATE INDEX ix_notifications_template_id ON notifications (template_id)')
+
+        # Drop indexes notification_history
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_key_type')
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_api_key_id')
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_notification_status')
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_notification_type')
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_service_id')
+        op.execute('DROP INDEX IF EXISTS ix_notification_history_week_created')
```

Comment:
have you tested the nightly tasks? they'll use notification history for anything with 3 days retention or less (ie both govuk and nhs coronavirus sms), as they both update for last four days.

are we worried that they might be super slow after this?

Reply:
These indexes are not used for that query. We found that just deleting the unused indexes helps performance on the notification_history table.